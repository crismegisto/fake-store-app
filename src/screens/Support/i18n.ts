import i18n from 'config/i18n';

i18n.addResources('en', 'SUPPORT', {
  SUBMIT: 'Submit',
  ERROR_NAME: 'The name is required.',
  TITLE_SUGGESTIONS: 'Put your suggestions in the following field',
  ERROR_SUGGESTIONS: 'Please provide us with more information...',
  THANKS: 'Thanks for your support'
});
