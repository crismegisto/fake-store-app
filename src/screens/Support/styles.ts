import { StyleSheet } from 'react-native';
import { primary, secondary, tertiary } from 'src/constants/colors';

const DEFAULT_SIZE = 40;
const DEFAULT_RADIOUS = 4;

const basicInput = {
  height: DEFAULT_SIZE,
  marginVertical: 12,
  borderWidth: 2,
  borderColor: primary,
  borderRadius: DEFAULT_RADIOUS,
  backgroundColor: 'white',
  padding: 10
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 10,
    justifyContent: 'center',
    backgroundColor: tertiary
  },
  button: {
    marginTop: 40,
    height: DEFAULT_SIZE,
    width: 120,
    alignItems: 'center',
    justifyContent: 'center',
    backgroundColor: secondary,
    borderRadius: DEFAULT_RADIOUS,
    alignSelf: 'center'
  },
  input: basicInput,
  largeInput: {
    ...basicInput,
    height: 200
  },
  suggestionsTitle: {
    marginTop: 14,
    marginHorizontal: 4
  },
  textError: {
    color: 'red'
  },
  textButton: {
    color: 'white',
    fontWeight: 'bold'
  },
  thanks: {
    color: primary,
    fontWeight: 'bold',
    fontSize: 16,
    alignSelf: 'center',
    marginTop: -50
  }
});

export default styles;
