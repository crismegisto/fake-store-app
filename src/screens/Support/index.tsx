import React from 'react';
import { Text, View, TextInput, TouchableOpacity } from 'react-native';
import { useForm, Controller } from 'react-hook-form';
import Animated, { withSpring, useSharedValue, useAnimatedStyle } from 'react-native-reanimated';
import { SuggestionsFormValues } from 'interfaces/suggestionsForm';
import { useTranslation } from 'react-i18next';

import './i18n';
import styles from './styles';

function Support() {
  const {
    control,
    handleSubmit,
    formState: { errors },
    reset
  } = useForm<SuggestionsFormValues>({
    defaultValues: {
      suggestions: 'Hello!\n'
    }
  });

  const { t } = useTranslation('SUPPORT');

  const offset = useSharedValue(1);

  const animatedStyles = useAnimatedStyle(() => ({
    transform: [
      {
        scale: withSpring(offset.value)
      }
    ]
  }));

  const onSubmit = () => {
    reset();
    offset.value = 0;
  };

  return (
    <View style={styles.container}>
      <Controller
        control={control}
        rules={{
          required: true
        }}
        render={({ field: { onChange, onBlur, value } }) => (
          <TextInput
            style={styles.input}
            placeholder="Name"
            maxLength={40}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
          />
        )}
        name="name"
      />
      {errors.name && <Text style={styles.textError}>{t('ERROR_NAME')}</Text>}

      <Controller
        control={control}
        render={({ field: { onChange, onBlur, value } }) => (
          <TextInput
            style={styles.input}
            placeholder="Email"
            maxLength={60}
            onBlur={onBlur}
            onChangeText={onChange}
            value={value}
          />
        )}
        name="email"
      />

      <Text style={styles.suggestionsTitle}>{t('TITLE_SUGGESTIONS')}</Text>
      <Controller
        control={control}
        rules={{
          minLength: 6
        }}
        render={({ field: { onChange, onBlur, value } }) => (
          <TextInput
            style={styles.largeInput}
            onBlur={onBlur}
            multiline
            onChangeText={onChange}
            testID="suggestionsForm"
            value={value}
          />
        )}
        name="suggestions"
      />
      {errors.suggestions && <Text style={styles.textError}>{t('ERROR_SUGGESTIONS')}</Text>}

      <Animated.View style={[animatedStyles]}>
        <TouchableOpacity style={styles.button} onPress={handleSubmit(onSubmit)}>
          <Text style={styles.textButton}>{t('SUBMIT')}</Text>
        </TouchableOpacity>
      </Animated.View>

      {offset.value === 0 && <Text style={styles.thanks}>{t('THANKS')}</Text>}
    </View>
  );
}

export default Support;
