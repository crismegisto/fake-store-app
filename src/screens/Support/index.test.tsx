import React from 'react';
import { screen, fireEvent, render } from '@testing-library/react-native';
import { act } from 'react-test-renderer';

import Support from './index';

beforeEach(() => {
  render(<Support />);
});

const name = 'Cristian';

describe('Support screen Tests', () => {
  test('match snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });

  test('textinputs work', () => {
    const email = 'cristian@gmail.com';
    const suggestions = 'Thank you!';

    const nameForm = screen.getByPlaceholderText('Name');
    const emailForm = screen.getByPlaceholderText('Email');
    const suggestionsForm = screen.getByTestId('suggestionsForm');

    expect(nameForm).toBeOnTheScreen();
    expect(emailForm).toBeOnTheScreen();
    expect(suggestionsForm).toBeOnTheScreen();

    fireEvent.changeText(nameForm, name);
    fireEvent.changeText(emailForm, email);
    fireEvent.changeText(suggestionsForm, suggestions);

    expect(nameForm.props.value).toBe(name);
    expect(emailForm.props.value).toBe(email);
    expect(suggestionsForm.props.value).toBe(suggestions);
  });

  test('Submit form without requirement', async () => {
    const submit = screen.getByText('Submit');
    expect(submit).toBeOnTheScreen();

    await act(() => {
      fireEvent.press(submit);
    });

    const errorName = screen.getByText('The name is required.');
    expect(errorName).toBeOnTheScreen();
  });
});
