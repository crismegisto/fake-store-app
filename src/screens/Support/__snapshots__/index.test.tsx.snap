// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Support screen Tests match snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#fefae0",
      "flex": 1,
      "justifyContent": "center",
      "padding": 10,
    }
  }
>
  <TextInput
    maxLength={40}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Name"
    style={
      {
        "backgroundColor": "white",
        "borderColor": "#606c38",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 40,
        "marginVertical": 12,
        "padding": 10,
      }
    }
  />
  <TextInput
    maxLength={60}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="Email"
    style={
      {
        "backgroundColor": "white",
        "borderColor": "#606c38",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 40,
        "marginVertical": 12,
        "padding": 10,
      }
    }
  />
  <Text
    style={
      {
        "marginHorizontal": 4,
        "marginTop": 14,
      }
    }
  >
    Put your suggestions in the following field
  </Text>
  <TextInput
    multiline={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": "white",
        "borderColor": "#606c38",
        "borderRadius": 4,
        "borderWidth": 2,
        "height": 200,
        "marginVertical": 12,
        "padding": 10,
      }
    }
    testID="suggestionsForm"
    value="Hello!
"
  />
  <View
    style={
      [
        {
          "transform": [
            {
              "scale": 1,
            },
          ],
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#283618",
          "borderRadius": 4,
          "height": 40,
          "justifyContent": "center",
          "marginTop": 40,
          "opacity": 1,
          "width": 120,
        }
      }
    >
      <Text
        style={
          {
            "color": "white",
            "fontWeight": "bold",
          }
        }
      >
        Submit
      </Text>
    </View>
  </View>
</View>
`;
